595/688 OsmChangesetSqlFileWriter/PostgresqlDumpfileWriter bounds writing

bwitham opened the pull request:
@mschicker Fairly simple one whenever you have the time.  I changed two OSM changeset writing classes that weren't previously writing the geospatial bounds of the changeset to do so.

mschicker commented on hoot-core/src/main/cpp/hoot/core/io/ApiDb.h:
```diff
@@ -190,6 +190,18 @@ class ApiDb
   QSqlDatabase getDB() { return _db; }
 
+  static long round(double x);
+  static long round(double x, int precision);
```
Could you add an explanation of the precision parameter in a comment? Also, my eyeball compiler thinks you'll see a divide-by-zero in this function if precision=1... that might be worth a note in a comment too.

bwitham replied:
We have that code sprinkled all over hoot but no good explanation for it (maybe some explanation in the docs).  It basically has to do with the storage precision differences between the hootenanny osm database and the public one.  I'll add some comments.

mschicker commented on hoot-core/src/main/cpp/hoot/core/io/ApiDb.cpp:
```diff
@@ -384,4 +401,14 @@ QSqlQuery ApiDb::_execNoPrepare(const QString& sql) const
+
+long ApiDb::round(double x, int precision)
+{
+  return (long)(floor(x * (10 * (precision - 1)) + 0.5) / (10 * (precision - 1)));
```
I'm looking at this, and... I'm thinking it's wrong. I think 10 should be raised to the power of (precision-1), not simply multiplied by (precision-1). Like use pow(10, precision-1) or something.

bwitham replied:
I'll take a closer look at it.  That was old, pre-existing code that I just refactored into ApiDb....just cut and paste and didn't really look at it too closely :-)

mschicker commented:
@bwitham, It looks good! Let me know what you think about that round() function (I might just be misunderstanding something), and then when that's resolved, I'll merge the pull!

bwitham commented:
@mschicker Thanks for looking at it.  I'll take a look at the rounding next week.

bwitham commented:
@mschicker It looks like both the original rounding function and the one changed in the way you described both yield the same result when used to convert between hoot api and osm api db coords.  See the test I added: ApiDbTest.  Also added the divide by zero comment.

mschicker commented on hoot-core-test/src/test/cpp/hoot/core/io/ApiDbTest.cpp:
```diff
@@ -0,0 +1,78 @@
+  {
+    const double hootApiCoordValue = 38.8543319;
+    long osmApiDbCoordValue = (long)ApiDb::round(hootApiCoordValue * ApiDb::COORDINATE_SCALE, 7);
+    CPPUNIT_ASSERT_EQUAL(osmApiDbCoordValue, (long)388543319);
```
This assert shows that the call to round isn't really doing anything. If we were truly rounding to 7 digits, we should assert against 388543300.

bwitham replied:
You're right, its verifying it at nine significant digits and not seven....so the test doesn't really mean much.

mschicker commented on hoot-core-test/src/test/cpp/hoot/core/io/ApiDbTest.cpp:
```diff
@@ -0,0 +1,78 @@
+  {
+    const double hootApiCoordValue = 38.8543319;
+    long osmApiDbCoordValue = (long)ApiDb::round2(hootApiCoordValue * ApiDb::COORDINATE_SCALE, 7);
+    CPPUNIT_ASSERT_EQUAL(osmApiDbCoordValue, (long)388543319);
```
Same deal here... round2 will round to the specified digit AFTER the decimal place - but that test thrown away anyway with the conversion to long.

mschicker commented:
@bwitham, I don't think the rounding is quite being used as intended - but it doesn't seem to be causing a problem, and probably isn't a good reason to hold up this pull-request any longer. What do you think?

bwitham commented:
@mschicker I think it is ok to merge at this point, since that rounding code has been used in that way for a few years or so.  BUT since it isn't completely intuitive why its being used as part of the coordinate conversion, at the very leas itt deserves either some more commenting or maybe a rewrite of the coord conversion code lines themselves.

@jasonsurratt When I started doing this osmapi db work, I was just copy/pasting the existing coord conversion code that uses the ApiDb::round function here and not giving much thought to what was actually going on.  Do you think "round" a bad name for this method, or is the conversion code I'm using just taking a shortcut using that method plus the truncation at the end to get the desired result, and that fact maybe needs to be made more clear or rewritten inside of another new method to be more clear?

bwitham commented:
working on getting rid of the round function with the precision input param now.

mschicker commented:
I think that's a good call :+1:

bwitham commented:
Thanks for the help, guys.  Latest changes should make more sense for coordinate conversion.

jasonsurratt commented:
Much cleaner.

mschicker commented:
Looks good. Thanks for slogging through the rounding stuff!

bwitham commented:
Thanks, man.  That's what the reviews are all about...getting it as right early on as possible.  As this MapEdit conflation task passes the 2 month mark, I'm hurrying to get it all working but the review keeps me in check :-)
